Engineering/Frontend/Prevent Layout Thrashing.md

```diff
@@ -0,0 +1,84 @@
+---
+tags: engineering/frontend,web-performance
+author: Le Minh Thanh
+date: 2022-09-11
+---
+### What is Layout thrashing
+![[layout-thrashing.png]]
+
+Layout thrashing means Forcing the browser to calculate a layout that is never rendered to the screen, and it hurt your performance so bad
+
+### Why
+Layout Thrashing happens, when you request layout information of an element or the document, while layout is in an **invalidated state**.
+
+```js
+// any DOM or CSSOM change flags the layout as invalid
+document.body.classList.add('foo');
+
+// reads layout == forces layout calculation
+const box = element.getBoundingClientRect();
+
+// write/mutate
+document.body.appendChild(someBox);
+
+//read/measure
+const color = getComputedStyle(someOtherBox).color;
+```
+
+Mixing Layout Read & Layout Mutation needs to wait for browser to recalculate layout & reflow to return you Layout value
+
+![[dont-touch-me.png]]
+
+### How to fix
+SEPARATE YOUR READS FROM YOUR WRITES AND THE PROBLEM IS SOLVED
+
+- Batch Read Layout first
```

Comment:
I remember we can use [Fragment](https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment) to batch changes :)

Reply:
> Great insight about performance. Not sure if it's ready to review?

Please do, I think it's ready to go

> I remember we can use [Fragment](https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment) to batch changes :)

Great, this is the first time I heard abt it. It seems promising for adding bunch of DOM nodes but still not ready to use, I guess

> The performance benefit of DocumentFragment is often overstated. In fact, in some engines, using a DocumentFragment is slower than appending to the document in a loop as demonstrated
-- From their doc